compiler/rustc_index/src/bit_set.rs

```diff
@@ -651,24 +651,174 @@ impl<T: Idx> BitRelations<ChunkedBitSet<T>> for ChunkedBitSet<T> {
 
 impl<T: Idx> BitRelations<HybridBitSet<T>> for ChunkedBitSet<T> {
     fn union(&mut self, other: &HybridBitSet<T>) -> bool {
-        // FIXME: this is slow if `other` is dense, and could easily be
-        // improved, but it hasn't been a problem in practice so far.
         assert_eq!(self.domain_size, other.domain_size());
-        sequential_update(|elem| self.insert(elem), other.iter())
+
+        match other {
+            HybridBitSet::Sparse(_) => sequential_update(|elem| self.insert(elem), other.iter()),
+            HybridBitSet::Dense(dense) => self.union(dense),
+        }
     }
 
     fn subtract(&mut self, other: &HybridBitSet<T>) -> bool {
-        // FIXME: this is slow if `other` is dense, and could easily be
-        // improved, but it hasn't been a problem in practice so far.
         assert_eq!(self.domain_size, other.domain_size());
-        sequential_update(|elem| self.remove(elem), other.iter())
+
+        match other {
+            HybridBitSet::Sparse(_) => sequential_update(|elem| self.remove(elem), other.iter()),
+            HybridBitSet::Dense(dense) => self.subtract(dense),
+        }
     }
 
     fn intersect(&mut self, _other: &HybridBitSet<T>) -> bool {
         unimplemented!("implement if/when necessary");
     }
 }
 
+impl<T: Idx> BitRelations<BitSet<T>> for ChunkedBitSet<T> {
+    fn union(&mut self, other: &BitSet<T>) -> bool {
+        assert_eq!(self.domain_size, other.domain_size());
+
+        let mut changed = false;
+        for (chunk, other_words) in self.chunks.iter_mut().zip(other.words().chunks(CHUNK_WORDS)) {
```

Comment:
Where is this `chunks()` method defined?

Reply:
https://doc.rust-lang.org/std/primitive.slice.html#method.chunks :+1: